element-view/src/androidMain/kotlin/ElementView.kt

```diff
@@ -0,0 +1,140 @@
+        super.onAttachedToWindow()
+        adapter?.onAttached(this)
+    }
+
+    override fun onDetachedFromWindow() {
+        super.onDetachedFromWindow()
+        adapter?.onDetached()
+    }
+
+    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
+        adapter?.onSizeChanged(w, h)
+    }
+
+    override fun onDraw(canvas: Canvas) {
+        val bitmap = adapter?.bitmap
+        if (bitmap != null) {
+            canvas.drawBitmap(bitmap, 0f, 0f, blitPaint)
+        }
+    }
+
+    public class Adapter<T>(
+        private val dataSource: Flow<T>,
+        private val updater: Updater<T>,
+    ) {
+
+        /** Android's Size was added in API 21, so do this to avoid bumping min version. */
+        public data class State(val view: ElementView?, val root: RootElement, val width: Int, val height: Int)
+
+        public fun interface Updater<T> {
+            public fun update(root: RootElement, width: Float, height: Float, data: T)
+        }
+
+        private val state: MutableStateFlow<State> = MutableStateFlow(EMPTY_STATE)
+
+        /** Coroutine scope rendering occurs on. This scope is live between calls to [onAttached] and [onDetached]. */
+        private var renderScope: CoroutineScope? = null
+            set(value) {
+                if (field !== value) field?.cancel()
+                field = value
+            }
+
+        /** The most recent bitmap to finish rendering, if any. */
+        internal var bitmap: Bitmap? = null
+            private set
+
+        /**
+         * The view changed size, so we should re-render the bitmap. When this happens, the RootElement
+         * will be reset, so the render will have a clean slate.
+         */
+        internal fun onSizeChanged(width: Int, height: Int) {
+            state.value = state.value.copy(root = RootElement(), width = width, height = height)
+        }
+
+        /**
+         * Attach this to a view (which is itself attached to a window).
+         * This will immediately render the current state if possible.
+         */
+        internal fun onAttached(view: ElementView) {
+            state.value = State(view, RootElement(), view.width, view.height)
+            renderScope = object : CoroutineScope {
+                private val job = Job()
+                override val coroutineContext: CoroutineContext get() = job
+            }.also { launchRenderingIn(it) }
+        }
+
+        /** Detach this from a view (usually because the view was detached from a window). */
+        internal fun onDetached() {
+            renderScope = null
+            state.value = EMPTY_STATE
+        }
+
+        private fun launchRenderingIn(scope: CoroutineScope) = scope.launch {
+            state.collectLatest { state ->
+                if (state.view == null) return@collectLatest
+                if (state.width == 0 || state.height == 0) return@collectLatest
+                dataSource.conflate().collect { data ->
```

Comment:
have we tested how performant this is on a couple different phones?

Reply:
Added simple fps logic in a local branch, `System.currentTimeMillis()` to measure time from `invalidate` call to next bitmap finishing drawing. 

Recorded this on my worst/oldest test device, a Samsung Galaxy A5 (released 2014). Frame rate isn't super smooth, but since this is off the main thread if you had other animations going they wouldn't be interrupted. Worst frame rendered in 32 milliseconds (> 30 fps).

https://user-images.githubusercontent.com/20380774/144545915-2e13f1dc-6943-4308-a493-d6a5569255e9.mp4

Reply:
After going through the effort of benchmarking that, I'm realizing that the current method of updating the bitmap directly inside the `collect` is basically a bad implementation of vsync. I'll look into a better method, or worst case just add a TODO.

Reply:
Frame rates should be fixed after the latest commit. We still drop the occasional frame on this old Samsung (2014!), but it's now butter smooth on my more modern devices.

https://user-images.githubusercontent.com/20380774/144551447-411bb502-ba61-4f87-b296-2946a2317dba.mp4